File: Python/stack/stack.py
class StackNode:
    def __init__(self, data):
        self.data = data
        self.next = None


class Stack:
    def __init__(self, elems = None):
        self._top = None
        self._length = 0
        if elems:
            for e in elems:
                self.push(e)
    
    def __len__(self):
        return self._length
    
    def __contains__(self, e):
        return self.contains(e)
    
    def empty(self):
        return self._length == 0
    
    def push(self, e):
        if self._length == 0:
            self._top = StackNode(e)
        else:
            temp = StackNode(e)
            temp.next = self._top
            self._top = temp
        self._length += 1
    
    def peek(self):
        if self._top == None:
            return None
        return self._top.data
    
    def pop(self):
        if self._length == 0:
            return None
        data = self._top.data
        self._top = self._top.next
        self._length -= 1
        return data
    
    def contains(self, e):
        current = self._top
        while current:
            if current.data == e:
                return True
            current = current.next
        return False
    
    def clear(self):
        self._top = None
        self._length = 0

File: JavaScript/stack/stack.js
class StackNode {
	constructor(data) {
		this.data = data;
		this.next = null;
	}
}

class Stack {
	constructor(elems = null) {
		this._top = null;
		this._length = 0;
		if (elems) {
			for (const e of elems) {
				this.push(e);
			}
		}
	}
	
	get length() {
		return this._length;
	}
	
	isEmpty() {
		return this._length == 0;
	}
	
	push(e) {
		if (this._length == 0) {
			this._top = new StackNode(e);
		}
		else {
			let temp = new StackNode(e);
			temp.next = this._top;
			this._top = temp;
		}
		this._length++;
	}
	
	peek() {
		if (this._top == null) {
			return null;
		}
		return this._top.data;
	}
	
	pop() {
		if (this._length == 0) {
			return null;
		}
		let data = this._top.data;
		this._top = this._top.next;
		this._length--;
		return data;
	}
	
	contains(e) {
		let current = this._top;
		while (current != null) {
			if (current.data == e) {
				return true;
			}
			current = current.next;
		}
		return false;
	}
	
	clear() {
		this._top = null;
		this._length = 0;
	}
}